rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,527 @@
+
+1. All document transforms need to be deterministic, that is a document
+   transform will always return the same result for the same set of inputs. 
+2. It should always be possible to construct the exact set of inputs required
+   for (1) at any point during the migration process (before, during, after). 
+
+Although these restrictions require significant changes, it does not prevent
+known upcoming migrations such as [sharing saved-objects in multiple spaces](https://github.com/elastic/kibana/issues/27004) or [splitting a saved
+object into multiple child
+documents](https://github.com/elastic/kibana/issues/26602). To ensure that
+these migrations are idempotent, they will have to generate new saved object
+id's deterministically with e.g. UUIDv5.
+
+### 4.3.1.2 Algorithm
+
+1. On startup, find the index to which the `.kibana` alias points,
+   `.kibana_n`, and check if any documents have an outdated `migrationVersion`
+   indicating that a migration needs to be performed. 
+2. Create `.kibana_n+1`, if it already exists, ignore the error and continue.
+3. Migrate documents by reading from `.kibana_n` and creating documents in
+   `.kibana_n+1`. If a document already exists, ignore.
+4. Move the `.kibana` index to point to `.kibana_n+1` (ignore if it has
+   already been moved). 
+
+Together with the limitations this algorithm ensures that migrations are
+idempotent. If two nodes are started simultaneously, both of them will start
+writing into `.kibana_n+1` but because migrations are idempotent, it doesn’t
+matter which node’s writes win.
+
+> Note: This algorithm doesn't support mixed Kibana versions attempting to
+> perform a migration at the same time. E.g. if a v7.8.0 and a v7.9.0 node
+> attempts to simultaneously migrate a 7.6.0 index the result will be an
+> inconsistent state. This is an unlikely scenario in practise and therefore
+> an acceptable limitation. In 8.x we can ensure that only the latest node
+> completes the migration. 
+
+## 4.5 Changes in 8.0
+1. Migrations will be in-place and re-use the same index for minor and patch
+   upgrades. 
+2. Only allow breaking field mapping changes (which requires creating a new
+   index) in a major.
+3. All update operations should use optimistic concurrency control.
+
+### 4.5.1 Migration algorithm (8.0):
+
+1. On startup, find the index to which the `.kibana` alias points,
+   `.kibana_n`, and check if any documents have an outdated `migrationVersion`
+   indicating that a migration needs to be performed.
+2. If there are documents with newer `migrationVersion` numbers, exit Kibana
+   with a fatal error.
+3. If there are outdated documents, migrate these in batches:
+   1. Read a batch of outdated documents from `.kibana_n`.
+   2. Transform documents by applying the migration functions.
+   3. Update the document batch in the same index using optimistic concurrency
+      control. 
+   4. If a batch fails due to an update version mismatch, repeat the entire
```

Comment:
In what circumstances would a batch fail? Writes are locked during this period, right? But maybe another node started updating this batch already? But if multiple nodes can run through this, it seems like the second node might hit `2.` because there will now be documents with newer `migrationVersion` numbers?

Reply:
A batch will fail (with an optimistic concurrency version mismatch) whenever a write has occurred after the batch was read. This can happen in the following scenarios:
 a. an outdated Kibana continues to write to the index (we try to minimize this in 4.5.2, but it's still possible)
 b. another Kibana node of the same version is also performing the migration in parallel
 c. a newer Kibana node is performing the migration

Step (2)'s "newer migrationVersion numbers" refers to documents with `migrationVersion > kibanaVersion` caused by (c). So if we hit scenario (b), we won't fail, we will just check if the migration is complete, and then read a new batch of documents that could be migrated.